src/search/command_line.cc

```diff
@@ -184,6 +184,118 @@ shared_ptr<SearchAlgorithm> parse_cmd_line(
+    const string &last_arg = parsed_args.back();
+    vector<string> suggestions;
+    if (find(parsed_args.begin(), parsed_args.end(), "--help") != parsed_args.end()) {
+        suggestions.push_back("--txt2tags");
+        plugins::Registry registry = plugins::RawRegistry::instance()->construct_registry();
+        for (const shared_ptr<const plugins::Feature> &feature : registry.get_features()) {
+            suggestions.push_back(feature->get_key());
+        }
+    } else if (last_arg == "--internal-plan-file") {
+        /* Suggest filename starting with current_word.
+           Handeled by default bash completion. */
+        exit(1);
+    } else if (last_arg == "--internal-previous-portfolio-plans") {
+        /* We want no suggestions and expect an integer
+           but we cannot avoid the default bash completion. */
+    } else if (last_arg == "--search") {
+        /* Return suggestions for the search string based on current_word.
+           Not implemented at the moment. */
+        exit(1);
+    } else {
+        // not completing an argument
+        suggestions.push_back("--help");
+        suggestions.push_back("--search");
+        suggestions.push_back("--internal-plan-file");
+        suggestions.push_back("--internal-previous-portfolio-plans");
+        suggestions.push_back("--if-unit-cost");
+        suggestions.push_back("--if-non-unit-cost");
+        suggestions.push_back("--always");
+        // remove suggestions not starting with current_word
+    }
+
+    if (!current_word.empty()) {
+        // Suggest only words that match with current_word
+        suggestions.erase(
+            remove_if(suggestions.begin(), suggestions.end(),
+                      [&](const string &value) {
+                          return !value.starts_with(current_word);
+                      }), suggestions.end());
+    }
+    return suggestions;
+}
+
+void handle_tab_completion(int argc, const char **argv) {
+    if (argc < 2 || static_cast<string>(argv[1]) != "--bash-complete") {
+        return;
+    }
+    if (argc < 5) {
+        input_error(
+            "The option --bash-complete is only meant to be called "
+            "internally to generate suggestions for tab completion.\n"
+            "Usage:\n    ./downward --bash-complete "
+            "$COMP_POINT \"$COMP_LINE\" $COMP_CWORD ${COMP_WORDS[@]}\n"
+            "where the environment variables have their usual meaning for bash completion:\n"
+            "$COMP_POINT is the position of the cursor in the command line.\n"
+            "$COMP_LINE is the current command line.\n"
+            "$COMP_CWORD is an index into ${COMP_WORDS} of the word under the cursor.\n"
+            "$COMP_WORDS is the current command line split into words.\n"
+            );
+    }
+    int cursor_pos = parse_int_arg("COMP_POINT", static_cast<string>(argv[2]));
+    const string command_line = static_cast<string>(argv[3]);
```

Comment:
We generally don't use `const` for local variables like these (e.g. we don't use `const int` in the line above).